### PR TITLE
Madrid

### PR DESCRIPTION
**Pull request by sheriffff**
Minor fix in `barcelona` and addition of `madrid`

### CODE REVIEW

File: madrid/bicimad.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    @shout
+    def get_stations_dataframe(self):
+        n_stations, datetime, data = self._get_stations_raw_info()
+
+        df = pd.DataFrame(data)[self.stations_keys]
+
+        df['datetime'] = pd.to_datetime(datetime)
+
+        return df
```
Comment (cimentadaj) [inline madrid/bicimad.py]: This is just an observation but do we really need to convert the JSON to a `pd.DataFrame`? I feel that's a step that could be avoided and no need to import the `pandas` dependency. In the end, we're just gonna push the JSON in someway to the database. I guess `pandas` will be needed when we export it from database to do some type of analysis.

File: madrid/bicimad.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
+import requests
+import pandas as pd
+
+from madrid.auth_credentials import auth_bicimad
```
Comment (cimentadaj) [inline madrid/bicimad.py:2]: As you say, when need to probably put this as an environment variable but it's ok for now.

File: madrid/bicimad.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    """
+    Class in charge of connecting to BiciMad API and querying information
+    Documentation here: https://apidocs.emtmadrid.es/
+    """
+    url_login = 'https://openapi.emtmadrid.es/v1/mobilitylabs/user/login/'
+    url_logout = 'https://openapi.emtmadrid.es/v1/mobilitylabs/user/logout/'
+    url_who_am_i = 'https://openapi.emtmadrid.es/v1/mobilitylabs/user/whoami/'
+    url_get_all_stations_info = 'https://openapi.emtmadrid.es/v1/transport/bicimad/stations/'
+
+    stations_keys = ['id', 'number', 'activate', 'total_bases', 'dock_bikes', 'free_bases', 'reservations_count']
```
Comment (cimentadaj) [inline madrid/bicimad.py:23]: Are these all the available columns or are there more? For example, for bcn there is a column which shows the five closest stations to each particular station. Just wondering whether you selected only these or whether there are more.

**Review by cimentadaj — CHANGES_REQUESTED** [review]
I really like your approach, really good job! The only thing I expect to be argued is whether we need to use `pandas` at this point or not, otherwise, it's approved.

File: madrid/bicimad.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    def __init__(self):
+        self._load_access_token()
+
+    @staticmethod
```
Comment (cimentadaj) [inline madrid/bicimad.py:28]: I tried reading about it, but can you tell me what `staticmethod` does?

**Comment by cimentadaj** [comment]
As soon as you take care of those, I'll send a PR with my changes from Barcelona. I think the next step is to setup a database on a server and write two equivalent methods for Barcelona/Madrid for pushing to the database. For this we need to talk in depth about how we're going to harmonize both datasets to be comparable.

**Comment by cimentadaj** [comment]
Can you also add a note on the `README` on what type of data is needed for the auth and where to place the file? So for example:

To reproduce the results for Madrid you'll need to get an account at XXXX and get XXX information. Once you have that, save these details in a file at `directory/where/auth/should/go` in this format:

```python
auth_bicimad = {
    'email': your email,
    'password': your password,
    'X-ClientId': string with X-ClientId,
    'X-ApiKey': string with X-ApiKey,
    'passKey': string with pass key
}
```

Also, I think we'll need to use pandas to clean the weather data so just ignore my pandas comments the I placed above.

File: madrid/bicimad.py (hunk 1 above)
Comment (sheriffff) [reply madrid/bicimad.py]: I totally agree. Let me change it.

File: madrid/bicimad.py (hunk 3 above)
Comment (sheriffff) [reply madrid/bicimad.py:23]: There are some more, which are in principle static (not changing with time: lat, lon), but as you say, why not save them in ddbb and just discard them in the future if not needed.

File: madrid/bicimad.py (hunk 4 above)
Comment (sheriffff) [reply madrid/bicimad.py:28]: It is a method that does not use `self`, so in principle could be defined out of the class, as an independent function. But it is left inside for keeping the code toguether.

**Pull request merged**